device/transport/amqp/src/amqp.ts

```diff
@@ -74,54 +83,331 @@ export class Amqp extends EventEmitter implements Client.Transport, StableConnec
-              }
-            });
+      if (eventName === 'message' && this.listeners('message').length === 0) {
+        debug('automatically detaching the link after the last message listener was removed');
+        this._fsm.handle('detachC2DLink', (err) => {
+          if (err) {
+            this.emit('errorReceived', err);
           }
-        } else {
-          debug('No active C2D link: nothing to do here.');
-        }
+        });
+      } else {
+        debug('removing listener for event: ' + eventName);
       }
     });
 
     /*Codes_SRS_NODE_DEVICE_AMQP_RECEIVER_16_003: [The `Amqp` object shall listen to the `message` and error events of the underlying `ReceiverLink` object when it has listeners on its `message` event.]*/
     this.on('newListener', (eventName, eventCallback) => {
       debug('registering new event handler for: ' + eventName);
       if (eventName === 'message') {
-        if (!this._c2dLink) {
-          this._amqp.attachReceiverLink(this._c2dEndpoint, null, (err, link) => {
-            this._c2dLink = link;
-            this._c2dLink.on('error', c2dErrorListener);
-            this._c2dLink.on('message', c2dMessageListener);
+        if (this.listeners('message').length === 0) {
+          debug('automatically attaching the link after the first message listener was registered');
+          this._fsm.handle('attachC2DLink', (err) => {
+            debug('C2D link attached automatically');
+            if (err) {
+              debug('error attaching the C2D link automatically: ' + err.toString());
+              this.emit('errorReceived', err);
+            }
           });
-        } else {
-          debug('c2d link already active: nothing to do here.');
         }
+      } else {
+        debug('adding listener for event: ' + eventName);
+      }
+    });
+
+    this._fsm = new machina.Fsm({
+      initialState: 'disconnected',
+      states: {
+        disconnected: {
+          _onEnter: (callback, err, result) => {
+            if (callback) {
+              if (err) {
+                callback(err);
+              } else {
+                callback(null, result);
+              }
+            }
+          },
+          connect: (connectCallback) => this._fsm.transition('connecting', connectCallback),
+          disconnect: (disconnectCallback) => {
+            if (disconnectCallback) {
+              disconnectCallback(null, new results.Disconnected());
+            }
+          },
+          sendEvent: (message, sendCallback) => {
+            /*Codes_SRS_NODE_DEVICE_AMQP_16_024: [The `sendEvent` method shall connect and authenticate the transport if necessary.]*/
+            this._fsm.handle('connect', (err, result) => {
+              if (err) {
+                sendCallback(err);
+              } else {
+                this._fsm.handle('sendEvent', message, sendCallback);
+              }
+            });
+          },
+          updateSharedAccessSignature: (token, callback) => {
```

Comment:
>updateSharedAccessSignature [](start = 10, length = 27)

naming feels wrong.  This isn't an action to update the signature.  This is an action that is taken after the signature was updated.  Maybe the generic handleSignatureUpdate?

Reply:
this is true, but we need to keep consistency for now. sadly wontfix.

---
In reply to: [141128283](https://github.com/Azure/azure-iot-sdk-node/pull/89#discussion_r141128283) [](ancestors = 141128283)